### PR TITLE
Development

### PR DESCRIPTION
**Pull request by RadiactiveJesus**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class User < ApplicationRecord
+  has_secure_password
+  validates :email, presence:true, uniqueness:true
+  validates :username, presence:true, uniqueness:true
+  validates :password, length: {minimum: 8}, allow_blank: false
+  before_create{generate_token(:remember_token)}
+  has_many :events, foreign_key: 'user_id', class_name: 'Event'
```
Comment (efrapp) [inline app/models/user.rb]: `[Optional]` Guys, if you are using the standard name (in this case `:events`), you don't need to specify the `:foreign_key` and `:class_name`

File: db/migrate/20190901150206_add_accepted_to_events_attendees.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AddAcceptedToEventsAttendees < ActiveRecord::Migration[5.2]
+  def change
+    add_column :events_attendees, :accepted, :boolean, default: false
+    add_column :events_attendees, :declined, :boolean, default: false
```
Comment (efrapp) [inline db/migrate/20190901150206_add_accepted_to_events_attendees.rb]: You don't need to change this but with just one field is enough. Just one boolean value (`accepted`, for example) can handle both states.

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    end
+  end
+
+  # GET /events/1
+  # GET /events/1.json
+  def show
+    @event = Event.find(params[:id])
+    @event_attendees = @event.attendees
+    #@going = @event.attendees.where('accepted AND NOT declined')
+    #@not_going = @event.attendees.where('NOT accepted AND declined')
```
Comment (efrapp) [inline app/controllers/events_controller.rb]: Please, remove code comments wherever they are in your project.

File: app/controllers/events_controller.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
+class EventsController < ApplicationController
+  before_action :set_event, only: [:show, :edit, :destroy, :attend]
+  #before_action :correct_user, only: %i[edit update]
```
Comment (efrapp) [inline app/controllers/events_controller.rb]: Remove this code comment if you don't need it.

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  def edit
+    session[:event_id] = params[:id]
+    @event = Event.find(params[:id])
+    @users = User.where('id != ?', current_user.id)
+  end
+
+  # POST /events
+  # POST /events.json
+  def create
+    @event = Event.new(event_params.merge(user_id: current_user.id))
```
Comment (efrapp) [inline app/controllers/events_controller.rb]: Guys, there is a simple way to do this using the `build` method generated by the relation between users and events. Please, change this line to use that method.

File: app/controllers/events_controller.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    respond_to do |format|
+      if @event.save
+        format.html { redirect_to @event, notice: 'Event was successfully created.' }
+        format.json { render :show, status: :created, location: @event }
+      else
+        format.html { render :new }
+        format.json { render json: @event.errors, status: :unprocessable_entity }
```
Comment (efrapp) [inline app/controllers/events_controller.rb]: I see that you use the scaffold to generate the actions but it is better if you create the methods manually because with the scaffold you have a lot of code that you don't need.

File: app/controllers/events_controller.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
+class EventsController < ApplicationController
+  before_action :set_event, only: [:show, :edit, :destroy, :attend]
+  #before_action :correct_user, only: %i[edit update]
+
+  # GET /events
+  # GET /events.json
+  def index
+    if params[:events].nil? || params[:events][:type] == 'upcoming'
```
Comment (efrapp) [inline app/controllers/events_controller.rb]: `[Optional]` I don't think this condition is necessary. You can display both types of events at the same time.

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  # GET /users
+  # GET /users.json
+  def index
+    @users = User.all
+  end
+
+  # GET /users/1
+  # GET /users/1.json
+  def show
+    @user = User.find(params[:id])
```
Comment (efrapp) [inline app/controllers/users_controller.rb]: Why don't you use the `current_user` you already created?

Comment (efrapp) [inline app/controllers/users_controller.rb]: Why don't you use the `current_user` you already created?

File: Gemfile [hunk 9]
```diff
@@ -39,6 +41,9 @@ gem 'bootsnap', '>= 1.1.0', require: false
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem 'rspec-rails', '~> 3.5'
+  gem 'database_cleaner'
+  gem 'factory_girl_rails'
```
Comment (mcrd25) [inline Gemfile]: [OPTIONAL]
factory_girll is deprecated. Please try and use [factory_bot](https://github.com/thoughtbot/factory_bot) in the future.

File: app/controllers/users_controller.rb (hunk 8 above)
Comment (mcrd25) [reply app/controllers/users_controller.rb]: [OPTIONAL]
Yes, please consider using `current_user`. This is an unnecessary query.

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  end
+
+  # Checks if a given user is the currently logged in user.
+  def current_user?(user)
+    user == current_user
+  end
+
+  # Checks to see if any users are logged in.
+  def logged_in?
```
Comment (mcrd25) [inline app/helpers/sessions_helper.rb]: I believe this is the only real candidate for this helper file.

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+module SessionsHelper
```
Comment (mcrd25) [inline app/helpers/sessions_helper.rb]: You seem to have application logic inside of your helpers. Helper methods ought to be view specific; the fact that you have to import the helpers in Application Controller for you to have access to those methods in controllers while you do not even use it in your views reiterates this. The views on the other hand, automatically have access to helper methods. 
So, what you can do is add all methods which are not views specific to your application controller directly. What you can also do if your method list gets long, you may use concerns or a lib folder to separate methods by function, etc.

Basically helpers = views, controller = controller. Why write methods in helpers only to have to include it in controller after. Helpers are automatically accessible in views. there is a reason helpers is outside the context of controller folder. You can, however, use concerns in your controllers.

Read [this article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) for a clearer explanation. 🙂

File: app/views/events/_event.html.erb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<li class='event'>
+    <%= link_to event.location, event%>
```
Comment (mcrd25) [inline app/views/events/_event.html.erb]: Should this not be the title of the event? 🤔

**Comment by RadiactiveJesus** [comment]
> 
> 
> Hi Team,
> **Great work making almost all the necessary corrections from the previous review**
> _Your RSPEC test all green now 😄_
> image
> 
> **Just one more thing to do**, I still can't identify the `previous events` from `upcoming events`and it was mentioned by the previous reviewer(The user's show page should not only show events a user has created but events a user has attended separated into an upcoming and past event. Utilizing the scope in your event model that you were to have created), checkout the screenshot from your web page;
> image
> 
> _Kindly re-submit for review after making this changes 🙂 Thumps -up await you....._

Me and my partner checked again what you are talking about and we find that in our computers is working as intended, can you please review it again? Or maay we are not understanding the problem that you are pointing out. Thanks

**Comment by bellom** [comment]
_In another word,_ The user's show page should not only show events a user has created but events a user has attended separated into an upcoming and past event. Utilizing the scope in your event model that you were to have created. Something like current_user.attended_events.prev etc.

That was listed from the previous reviewer, and that was what am pointing to.
We should have our `upcoming events` separated from `previous events`.
I hope that clear now.
